src/box/sql/mem.c

```diff
@@ -3284,6 +3284,71 @@ mem_encode_map(const struct Mem *mems, uint32_t count, uint32_t *size,
 	return NULL;
 }
 
+/* Locate an element in a MAP or ARRAY using the given key.*/
+static int
+mp_getitem(const char **data, const struct Mem *key)
+{
+	if (mp_typeof(**data) != MP_ARRAY && mp_typeof(**data) != MP_MAP) {
+		*data = NULL;
+		return 0;
+	}
+	const char *end = *data;
+	if (mp_typeof(**data) == MP_ARRAY) {
+		uint32_t size = mp_decode_array(data);
+		if (!mem_is_uint(key) || key->u.u == 0 || key->u.u > size) {
+			*data = NULL;
+			return 0;
+		}
+		for (uint32_t i = 0; i < key->u.u - 1; ++i)
+			mp_next(data);
+		return 0;
+	}
+	struct Mem mem;
+	mem_create(&mem);
+	uint32_t size = mp_decode_map(data);
+	for (uint32_t i = 0; i < size; ++i) {
+		uint32_t len;
+		if (mem_from_mp_ephemeral(&mem, *data, &len) != 0)
+			return -1;
+		assert(mem_is_trivial(&mem) && !mem_is_metatype(&mem));
+		*data += len;
+		if (mem_is_map(&mem) || mem_is_array(&mem)) {
+			*data = NULL;
+			return 0;
+		}
+		if (mem_cmp_scalar(&mem, key, NULL) == 0)
+			return 0;
+		mp_next(data);
+	}
+	mp_next(&end);
```

Comment:
Why do you need to decode it all second time? Below I added an assertion that `*data == end` always and the test passed (I didn't run all the tests though).

Reply:
You are right. I think at some point I used `break` to exit `for`. Fixed, removed `end` variable.